clippy_lints/src/no_effect.rs

```diff
@@ -130,10 +131,12 @@ fn has_no_effect(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
         },
         ExprKind::Call(callee, args) => {
             if let ExprKind::Path(ref qpath) = callee.kind {
-                let res = cx.qpath_res(qpath, callee.hir_id);
                 let def_matched = matches!(
-                    res,
-                    Res::Def(DefKind::Struct | DefKind::Variant | DefKind::Ctor(..), ..)
+                    cx.qpath_res(qpath, callee.hir_id),
+                    Res::Def(
+                        DefKind::Struct | DefKind::Variant | DefKind::Ctor(_, CtorKind::Fn | CtorKind::Fictive),
```

Comment:
I don't think this is the right fix. Perhaps try inspecting `tcx.type_dependent_def(callee.hir_id)`.

Add a test with
```rust
struct GreetStruct(u32);

impl FnOnce<(&str,)> for GreetStruct {
    type Output = ();

    extern "rust-call" fn call_once(self, (who,): (&str,)) -> Self::Output {
        println!("hello {}", who);
    }
}

fn main() {
    GreetStruct(1)("world");
}
```

Reply:
@camsteffen Thanks for suggestion, I am not that familiar with the rustc API. 
I looked into it briefly but it gives `none` as result, I think that it requires MethodCall and not Call as ExprKind.

Reply:
Hmm yeah I haven't seen it work with `Call` either, it was just a guess...

Reply:
Ah did you try `expr.hir_id`?

Reply:
@camsteffen I've tried it with expr.hir_id and it works. So one option would be to add something like this:
```rust
fn is_fn_once(cx: &LateContext<'_>, expr: &'a Expr<'a>) -> bool {
    if_chain! {
        if let Some(def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
        if match_def_path(cx, def_id, &paths::FN_ONCE_CALL_ONCE);
        then {
            return true;
        }
    }
    false
}
```
Hardcoded path is used as it seems `FnOnce` trait has no corresponding Symbol. But I've noticed we also have false positive for `FnMut` and probably `Fn` case, so those should also be added.

Other way that we could go is more similar to first solution I had, if we use:
`Res::Def(DefKind::Variant | DefKind::Ctor(_, CtorKind::Fn | CtorKind::Fictive), ..)` in match instead of `Res::Def(DefKind::Struct | DefKind::Variant | DefKind::Ctor(..), ..)` we avoid these cases. I think original intention was to cover construction of touples (as far as I can see from tests) but implementation seems to be too broad.

Reply:
I think you can just do this

```rust
ExprKind::Call(callee, args) => {
    if cx.typeck_results().type_dependent_def_id(expr.hir_id).is_some() {
        // type-dependent function call like `impl FnOnce for X`
        return false;
    }
    ..
```